[RHCLOUD-17247] Fetch tokens from marketplace and store them in an struct

MikelAlejoBR opened the pull request:
Implements a function that calls the marketplace endpoints to get a new token, and extracts the `access_token` and the `expiration` timestamp from the response, to return them in a `BearerToken` struct that we can use latter.

[[RHCLOUD-17247]](https://issues.redhat.com/browse/RHCLOUD-17247)

MikelAlejoBR committed:
feature: marketplace token fetching

MikelAlejoBR committed:
feature: read the marketplace's host from env variables

MikelAlejoBR committed:
refactor: add more tests

It adds tests for the token provider which check if errors are returned
in the following situations:

* The marketplace is unreachable.
* The marketplace returns a non 200 response.
* The marketplace returns a bad JSON.
* The marketpalce returns an unexpected JSON structure.

lindgrenj6 commented on marketplace/token.go:
```diff
@@ -0,0 +1,31 @@
+
+// DecodeMarketplaceTokenFromResponse decodes the bearer token and the expiration timestamp from the received
+// response.
+func DecodeMarketplaceTokenFromResponse(response *http.Response) (BearerToken, error) {
```
Can we change the return values to `(*BearerToken, error)` that way we can just `return nil, err`?

MikelAlejoBR replied:
Sure! Done.

lindgrenj6 commented on marketplace/marketplace_token_provider.go:
```diff
@@ -0,0 +1,47 @@
+}
+
+// GetToken sends a request to the marketplace to request a bearer token.
+func GetToken(httpClient httpClient, marketplaceHost string, apiKey string) (BearerToken, error) {
```
The change would bubble up here too - that way we dont have to pass around empty structs on error.

MikelAlejoBR replied:
Done as well!

lindgrenj6 commented on marketplace/marketplace_token_provider.go:
```diff
@@ -0,0 +1,47 @@
+}
+
+// GetToken sends a request to the marketplace to request a bearer token.
+func GetToken(httpClient httpClient, marketplaceHost string, apiKey string) (BearerToken, error) {
```
Since we're passing the interface in - should probably have a test for this too.

MikelAlejoBR replied:
Done too!

lindgrenj6 commented on marketplace/marketplace_token_provider.go:
```diff
@@ -0,0 +1,47 @@
+	request.Header.Add("Accept", "application/json")
+	request.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	response, err := httpClient.Do(request)
```
Don't forget to add a `defer response.Body.Close()` on the end here - otherwise the response leaks.

MikelAlejoBR replied:
Nice catch! I did a manual `close` on the decoding function — which I think that makes more sense, as it's the one opening the response's body as well — and logged the error to catch potential leaks.

lindgrenj6 reviewed (commented):
Just a few _super small_ changes - other than that looks good!

MikelAlejoBR committed:
refactor: make functions return a pointer to the token

That way the functions return "nil" in case of an error, instead of
allocating an empty "BearerToken" object.

MikelAlejoBR committed:
refactor: use pointers for the token's fields

Instead of checking against the default values of the primitive types,
this change lets checking against "nil" and making sure that, indeed,
the values were not present.

MikelAlejoBR committed:
refactor: grab the marketplace's host from the config

MikelAlejoBR committed:
fix: close response's body once we're done reading

MikelAlejoBR committed:
fix: create a "main_test" for the "marketplace" package

This is required to parse the "-createdb" and "-integration" flags even
if they're not used in the package, as otherwise the tests fail.

MikelAlejoBR committed:
refactor: remove forgotten "println"

lindgrenj6 commented on marketplace/marketplace_token_provider.go:
```diff
@@ -0,0 +1,49 @@
+
+// httpClient abstracts away the client to be used in the GetToken function, and allows mocking it easily for the
+// tests.
+type httpClient interface {
```
So one last request here - can we add a `var` that contains the instance of httpClient, so we can just set it once rather than passing it around as an argument. [Kind of like we do with the DAOs](https://github.com/RedHatInsights/sources-api-go/blob/dcc99ae23f9d1295a7ce77fb2167913e6f15322b/authentication_handlers.go#L13) except instead of a function we can just make it a field. You should be able to set that variable to your mocked out clients in the tests.

Set it to `http.Client{Timeout:10*time.Second}` in `main.go` (like we do with the daos), it's [concurrency safe](https://pkg.go.dev/net/http#Client) according to docs so we can just have one client that will be used by potentially multiple threads at once. Maybe in the future we can extract it to a util var or something - but I'm not sure we'll need an http client elsewhere except here. RBAC implements its own in the library for example.

MikelAlejoBR replied:
All righty. I did it, but I had to make the function to change the `var` exported, to be able to initialize it from the `main` function.

lindgrenj6 reviewed (commented):
Just one teeeeny change :pray:

MikelAlejoBR committed:
refactor: move "HttpClient" dependency to a variable

This makes it consistent with the way we have been injecting DAOs in our
application.

lindgrenj6 commented on marketplace/token.go:
```diff
@@ -0,0 +1,37 @@
+
+	err = response.Body.Close()
+	if err != nil {
+		log.Fatalf("could not close the marketplace JWT token response's body: %s", err)
```
I actually don't think we should fatal log here - maybe return an error instead and make that return a 500 response in the controller later?

MikelAlejoBR replied:
Whoops, you are right. I was intending to use `fmt.Errorf` there but I messed up. Nice catch!

lindgrenj6 commented on marketplace/token.go:
```diff
@@ -0,0 +1,37 @@
+import (
+	"encoding/json"
+	"fmt"
+	"log"
```
Should have caught this earlier - but we should be using the `logging` package instead of the std logger - that way it'll get thrown at cloudwatch and formatted as json.

MikelAlejoBR replied:
Yeah this is a mistake related to the previous one. I've removed this from here. I guess that we'll use `logging` extensively on the middleware or on the handlers to return user friendly messages, and log the error messages.

lindgrenj6 reviewed (commented):
Sorry just one more change that I should have noticed earlier but missed. We need to be using the logging package instead of the stdlib logger - that way it formats the logs correctly and sends them to cloudwatch.

MikelAlejoBR committed:
refactor: remove "Fatalf"

MikelAlejoBR committed:
refactor: use an interface for the provider

This will make injecting the dependency easier, and will make testing
and mocking this dependency easier.
